drivers/bluetooth/hci/userchan.c

```diff
@@ -93,6 +95,67 @@ static struct net_buf *get_rx(const uint8_t *buf)
 	return NULL;
 }
 
+/**
+ * @brief Decode the length of an HCI H4 packet
+ * @details Decodes packet length according to Bluetooth spec v5.4 Vol 4 Part E
+ * @param packet_buff	Pointer to a HCI packet buffer
+ * @return Length of the HCI packet in bytes, zero if no valid packet found.
+ */
+static uint16_t packet_len(const uint8_t *packet_buff)
+{
+	uint16_t payload_len = 0;
+	uint8_t header_len = 0;
+	const uint8_t packet_type = packet_buff[0];
+	const uint8_t *packet_hdr = &packet_buff[sizeof(packet_type)];
+
+	switch (packet_type) {
+	case H4_CMD: {
+		const struct bt_hci_cmd_hdr *hdr = (const struct bt_hci_cmd_hdr *)packet_hdr;
+
+		/* Parameter Total Length */
+		payload_len = hdr->param_len;
+		header_len = BT_HCI_CMD_HDR_SIZE;
+		break;
+	}
+	case H4_ACL: {
+		const struct bt_hci_acl_hdr *hdr = (const struct bt_hci_acl_hdr *)packet_hdr;
+
+		/* Data Total Length */
+		payload_len = sys_le16_to_cpu(hdr->len);
+		header_len = BT_HCI_ACL_HDR_SIZE;
+		break;
+	}
+	case H4_SCO: {
+		const struct bt_hci_sco_hdr *hdr = (const struct bt_hci_sco_hdr *)packet_hdr;
+
+		/* Data_Total_Length */
+		payload_len = hdr->len;
+		header_len = BT_HCI_SCO_HDR_SIZE;
+		break;
+	}
+	case H4_EVT: {
+		const struct bt_hci_evt_hdr *hdr = (const struct bt_hci_evt_hdr *)packet_hdr;
+
+		/* Parameter Total Length */
+		payload_len = hdr->len;
+		header_len = BT_HCI_EVT_HDR_SIZE;
+		break;
+	}
+	case H4_ISO: {
+		const struct bt_hci_iso_hdr *hdr = (const struct bt_hci_iso_hdr *)packet_hdr;
+
+		/* ISO_Data_Load_Length parameter */
+		payload_len =  bt_iso_hdr_len(sys_le16_to_cpu(hdr->len));
+		header_len = BT_HCI_ISO_HDR_SIZE;
+	}
+	/* If no valid packet type found */
+	default:
+		return 0;
```

Comment:
Might make sense to add a `LOG_WRN("Unknown packet type 0x%02x", type)` before the return.

Reply:
added